File: sdk/python/noise_server.py
#!/usr/bin/env python3
"""
Noise-NK TCP Server

A Python server that uses Noise-NK protocol for secure communication.
This server will accept connections from JavaScript clients and demonstrate
cross-platform compatibility.
"""

import sys
import os
import socket
import threading
import json
import time
from typing import Optional

# Add Python SDK to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'openadp'))

from openadp.client import NoiseNK, generate_keypair

class NoiseNKServer:
    """TCP server using Noise-NK protocol for secure communication."""
    
    def __init__(self, host='localhost', port=8888):
        self.host = host
        self.port = port
        self.server_socket = None
        self.running = False
        
        # Generate server static key pair
        self.server_private, self.server_public = generate_keypair()
        
        print(f"🔐 Server static public key: {self.server_public.hex()}")
        print(f"🔐 Clients should use this key to connect")
        
    def start(self):
        """Start the server."""
        self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        
        try:
            self.server_socket.bind((self.host, self.port))
            self.server_socket.listen(5)
            self.running = True
            
            print(f"🚀 Noise-NK server listening on {self.host}:{self.port}")
            print(f"📡 Waiting for JavaScript clients...")
            
            while self.running:
                try:
                    client_socket, address = self.server_socket.accept()
                    print(f"📞 New connection from {address}")
                    
                    # Handle client in a separate thread
                    client_thread = threading.Thread(
                        target=self.handle_client,
                        args=(client_socket, address)
                    )
                    client_thread.daemon = True
                    client_thread.start()
                    
                except socket.error as e:
                    if self.running:
                        print(f"❌ Socket error: {e}")
                    break
                    
        except Exception as e:
            print(f"❌ Server error: {e}")
        finally:
            self.stop()
    
    def handle_client(self, client_socket: socket.socket, address):
        """Handle a client connection with Noise-NK protocol."""
        try:
            print(f"🔒 Starting Noise-NK handshake with {address}")
            
            # Debug: Print the server's static key info
            print(f"🔑 Server private key: {self.server_private.hex()}")
            print(f"🔑 Server public key: {self.server_public.hex()}")
            
            # Initialize Noise-NK as responder
            noise = NoiseNK()
            noise.initialize_as_responder(self.server_private)
            
            # Debug: Check what key the noise instance is actually using
            try:
                # Let's see if we can get the static key from the noise instance
                if hasattr(noise.noise, 'handshake_state') and hasattr(noise.noise.handshake_state, 's'):
                    actual_key = noise.noise.handshake_state.s.public_bytes
                    print(f"🔍 Noise library is using static public key: {actual_key.hex()}")
                else:
                    print(f"🔍 Cannot inspect noise library's static key")
            except Exception as debug_e:
                print(f"🔍 Debug error: {debug_e}")
            
            # Receive first handshake message from client
            message1_data = self.receive_message(client_socket)
            if not message1_data:
                print(f"❌ Failed to receive handshake message 1 from {address}")
                return
                
            print(f"📨 Received handshake message 1: {len(message1_data)} bytes")
            print(f"🔍 Raw message 1 hex: {message1_data.hex()}")
            
            # Print handshake hash before reading first message (to compare with JS)
            try:
                hash_before = noise.get_handshake_hash()
                print(f"🔑 Python handshake hash BEFORE reading message 1: {hash_before.hex()}")
            except Exception as hash_e:
                print(f"⚠️ Could not get handshake hash before message 1: {hash_e}")
            
            # Process first handshake message
            try:
                payload1 = noise.read_message(message1_data)
                print(f"📝 Client payload 1: {payload1}")
                
                # Print handshake hash after reading first message
                try:
                    hash1 = noise.get_handshake_hash()
                    print(f"🔑 Python handshake hash AFTER reading message 1: {hash1.hex()}")
                except Exception as hash_e:
                    print(f"⚠️ Could not get handshake hash after message 1: {hash_e}")
            except Exception as e:
                print(f"❌ Failed to process handshake message 1: {e}")
                print(f"🔍 This might be due to payload encryption mismatch")
                import traceback
                traceback.print_exc()
                return
            
            # Send second handshake message (no payload to avoid AES-GCM issues)
            message2 = noise.write_message()
            
            if not self.send_message(client_socket, message2):
                print(f"❌ Failed to send handshake message 2 to {address}")
                return
                
            print(f"📤 Sent handshake message 2: {len(message2)} bytes")
            
            if not noise.handshake_complete:
                print(f"❌ Handshake not complete after message 2")
                return
                
            # Print final handshake hash
            try:
                final_hash = noise.get_handshake_hash()
                print(f"🔑 Python final handshake hash: {final_hash.hex()}")
            except Exception as hash_e:
                print(f"⚠️ Could not get final handshake hash: {hash_e}")
                
            print(f"✅ Noise-NK handshake completed with {address}")
            
            # Now handle secure messages
            self.handle_secure_communication(client_socket, address, noise)
            
        except Exception as e:
            print(f"❌ Error handling client {address}: {e}")
            import traceback
            traceback.print_exc()
        finally:
            client_socket.close()
            print(f"🔌 Disconnected from {address}")
    
    def handle_secure_communication(self, client_socket: socket.socket, address, noise: NoiseNK):
        """Handle secure communication after handshake completion."""
        print(f"🔐 Secure channel established with {address}")
        
        try:
            while True:
                # Receive encrypted message from client
                encrypted_data = self.receive_message(client_socket)
                if not encrypted_data:
                    print(f"📡 Client {address} disconnected")
                    break
                
                # Decrypt message
                try:
                    plaintext = noise.decrypt(encrypted_data)
                    if plaintext is None:
                        print(f"❌ Decryption returned None from {address}")
                        break
                    
                    message = plaintext.decode('utf-8')
                    print(f"📨 Received from {address}: {message}")
                    
                    # Send encrypted response
                    response = f"Echo: {message} (from Python server)"
                    encrypted_response = noise.encrypt(response.encode('utf-8'))
                    
                    if self.send_message(client_socket, encrypted_response):
                        print(f"📤 Sent response to {address}: {response}")
                    else:
                        print(f"❌ Failed to send response to {address}")
                        break
                        
                except Exception as e:
                    print(f"❌ Error processing encrypted message from {address}: {e}")
                    import traceback
                    traceback.print_exc()
                    break
                    
        except Exception as e:
            print(f"❌ Error in secure communication with {address}: {e}")
    
    def send_message(self, sock: socket.socket, data: bytes) -> bool:
        """Send a message with length prefix."""
        try:
            # Send length first (4 bytes, big-endian)
            length = len(data)
            length_bytes = length.to_bytes(4, 'big')
            sock.sendall(length_bytes)
            
            # Send data
            sock.sendall(data)
            return True
        except Exception as e:
            print(f"❌ Send error: {e}")
            return False
    
    def receive_message(self, sock: socket.socket) -> Optional[bytes]:
        """Receive a message with length prefix."""
        try:
            # Receive length first (4 bytes, big-endian)
            length_bytes = self.receive_exact(sock, 4)
            if not length_bytes:
                return None
                
            length = int.from_bytes(length_bytes, 'big')
            
            # Receive data
            data = self.receive_exact(sock, length)
            return data
            
        except Exception as e:
            print(f"❌ Receive error: {e}")
            return None
    
    def receive_exact(self, sock: socket.socket, length: int) -> Optional[bytes]:
        """Receive exactly the specified number of bytes."""
        data = b''
        while len(data) < length:
            chunk = sock.recv(length - len(data))
            if not chunk:
                return None
            data += chunk
        return data
    
    def stop(self):
        """Stop the server."""
        self.running = False
        if self.server_socket:
            self.server_socket.close()
        print("🛑 Server stopped")
    
    def get_server_info(self) -> dict:
        """Get server information for clients."""
        return {
            "host": self.host,
            "port": self.port,
            "public_key": self.server_public.hex(),
            "protocol": "Noise_NK_25519_AESGCM_SHA256"
        }

def main():
    """Run the Noise-NK server."""
    print("🔐 Noise-NK TCP Server")
    print("=====================")
    
    server = NoiseNKServer()
    
    # Save server info for clients
    server_info = server.get_server_info()
    with open('server_info.json', 'w') as f:
        json.dump(server_info, f, indent=2)
    
    print(f"💾 Server info saved to server_info.json")
    print(f"📋 Server info: {json.dumps(server_info, indent=2)}")
    
    try:
        server.start()
    except KeyboardInterrupt:
        print("\n🛑 Shutting down server...")
        server.stop()

if __name__ == "__main__":
    main() 

File: sdk/cpp/tools/noise_server.cpp
#include "openadp/noise.hpp"
#include "openadp/crypto.hpp"
#include "openadp/utils.hpp"
#include "openadp/debug.hpp"
#include <iostream>
#include <fstream>
#include <thread>
#include <chrono>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <signal.h>
#include <nlohmann/json.hpp>
#include <getopt.h>

using json = nlohmann::json;
using namespace openadp;

bool server_running = true;

void signal_handler(int signal) {
    std::cout << "\n🛑 Shutting down server..." << std::endl;
    server_running = false;
}

// Helper to send length-prefixed message
bool send_message(int socket_fd, const Bytes& data) {
    uint32_t length = htonl(data.size());
    
    // Send length
    if (send(socket_fd, &length, 4, 0) != 4) {
        return false;
    }
    
    // Send data
    if (send(socket_fd, data.data(), data.size(), 0) != static_cast<ssize_t>(data.size())) {
        return false;
    }
    
    return true;
}

// Helper to receive length-prefixed message
Bytes receive_message(int socket_fd) {
    uint32_t length;
    
    // Receive length
    if (recv(socket_fd, &length, 4, MSG_WAITALL) != 4) {
        return Bytes{};
    }
    
    length = ntohl(length);
    if (length > 1024 * 1024) { // 1MB limit
        return Bytes{};
    }
    
    // Receive data
    Bytes data(length);
    if (recv(socket_fd, data.data(), length, MSG_WAITALL) != static_cast<ssize_t>(length)) {
        return Bytes{};
    }
    
    return data;
}

void handle_client(int client_fd, const std::string& client_addr, const Bytes& server_private_key) {
    std::cout << "📞 New connection from " << client_addr << std::endl;
    std::cout << "🔒 Starting Noise-NK handshake with " << client_addr << std::endl;
    
    try {
        // Initialize Noise-NK server
        noise::NoiseState server;
        server.initialize_responder(server_private_key);
        
        // Receive first handshake message from client
        Bytes client_message = receive_message(client_fd);
        if (client_message.empty()) {
            std::cout << "❌ Failed to receive handshake message 1 from " << client_addr << std::endl;
            close(client_fd);
            return;
        }
        
        std::cout << "📨 Received handshake message 1: " << client_message.size() << " bytes" << std::endl;
        std::cout << "🔍 Raw message 1 hex: " << crypto::bytes_to_hex(client_message) << std::endl;
        
        // Process client handshake message
        Bytes client_payload = server.read_message(client_message);
        std::cout << "📝 Client payload 1: " << client_payload.size() << " bytes" << std::endl;
        
        // Send second handshake message to client
        Bytes server_message = server.write_message();
        std::cout << "📤 Sent handshake message 2: " << server_message.size() << " bytes" << std::endl;
        
        if (!send_message(client_fd, server_message)) {
            std::cout << "❌ Failed to send handshake message 2 to " << client_addr << std::endl;
            close(client_fd);
            return;
        }
        
        if (!server.handshake_finished()) {
            std::cout << "❌ Handshake not completed with " << client_addr << std::endl;
            close(client_fd);
            return;
        }
        
        std::cout << "🔑 C++ final handshake hash: " << crypto::bytes_to_hex(server.get_handshake_hash()) << std::endl;
        std::cout << "✅ Noise-NK handshake completed with " << client_addr << std::endl;
        std::cout << "🔐 Secure channel established with " << client_addr << std::endl;
        
        // Handle secure communication
        while (server_running) {
            Bytes encrypted_data = receive_message(client_fd);
            if (encrypted_data.empty()) {
                std::cout << "📡 Client " << client_addr << " disconnected" << std::endl;
                break;
            }
            
            try {
                // Decrypt message
                Bytes plaintext = server.decrypt(encrypted_data);
                std::string message = utils::bytes_to_string(plaintext);
                std::cout << "📨 Received from " << client_addr << ": " << message << std::endl;
                
                // Send echo response
                std::string response = "Echo: " + message + " (from C++ server)";
                Bytes response_bytes = utils::string_to_bytes(response);
                Bytes encrypted_response = server.encrypt(response_bytes);
                
                if (!send_message(client_fd, encrypted_response)) {
                    std::cout << "❌ Failed to send response to " << client_addr << std::endl;
                    break;
                }
                
                std::cout << "📤 Sent response to " << client_addr << ": " << response << std::endl;
                
            } catch (const OpenADPError& e) {
                std::cout << "❌ Error processing encrypted message from " << client_addr << ": " << e.what() << std::endl;
                break;
            }
        }
        
    } catch (const OpenADPError& e) {
        std::cout << "❌ Failed to process handshake message 1: " << e.what() << std::endl;
        std::cout << "🔍 This might be due to payload encryption mismatch" << std::endl;
    }
    
    close(client_fd);
    std::cout << "🔌 Disconnected from " << client_addr << std::endl;
}

void print_usage(const char* program_name) {
    std::cout << "Usage: " << program_name << " [OPTIONS]\n"
              << "\n"
              << "Simple Noise-NK server for testing OpenADP clients.\n"
              << "\n"
              << "Options:\n"
              << "  --port <num>          Port to listen on (default: 8080)\n"
              << "  --debug               Enable debug mode (deterministic operations)\n"
              << "  --help                Show this help message\n"
              << "\n"
              << "Examples:\n"
              << "  " << program_name << "\n"
              << "  " << program_name << " --port 9090\n";
}

int main(int argc, char* argv[]) {
    int port = 8080;
    bool debug_mode = false;
    
    static struct option long_options[] = {
        {"port", required_argument, 0, 'p'},
        {"debug", no_argument, 0, 'd'},
        {"help", no_argument, 0, 'h'},
        {0, 0, 0, 0}
    };
    
    int option_index = 0;
    int c;
    
    while ((c = getopt_long(argc, argv, "p:dh", long_options, &option_index)) != -1) {
        switch (c) {
            case 'p':
                port = std::stoi(optarg);
                break;
            case 'd':
                debug_mode = true;
                break;
            case 'h':
                print_usage(argv[0]);
                return 0;
            case '?':
                print_usage(argv[0]);
                return 1;
            default:
                break;
        }
    }
    
    // Set debug mode if requested
    if (debug_mode) {
        debug::set_debug(true);
    }
    
    std::cout << "🔐 Noise-NK C++ TCP Server" << std::endl;
    std::cout << "===========================" << std::endl;
    
    // Set up signal handler
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    try {
        // Generate server keys
        Bytes server_private_key = noise::generate_keypair_private();
        Bytes server_public_key = noise::derive_public_key(server_private_key);
        
        std::string server_public_hex = crypto::bytes_to_hex(server_public_key);
        std::cout << "🔐 Server static public key: " << server_public_hex << std::endl;
        std::cout << "🔐 Clients should use this key to connect" << std::endl;
        
        // Save server info to JSON file
        json server_info = {
            {"host", "localhost"},
            {"port", port},
            {"public_key", server_public_hex},
            {"protocol", "Noise_NK_25519_AESGCM_SHA256"}
        };
        
        std::ofstream server_info_file("server_info.json");
        server_info_file << server_info.dump(2) << std::endl;
        server_info_file.close();
        
        std::cout << "💾 Server info saved to server_info.json" << std::endl;
        std::cout << "📋 Server info: " << server_info.dump(2) << std::endl;
        
        // Create socket
        int server_fd = socket(AF_INET, SOCK_STREAM, 0);
        if (server_fd < 0) {
            std::cerr << "❌ Failed to create socket" << std::endl;
            return 1;
        }
        
        // Set socket options
        int opt = 1;
        if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)) < 0) {
            std::cerr << "❌ Failed to set socket options" << std::endl;
            close(server_fd);
            return 1;
        }
        
        // Bind socket
        struct sockaddr_in server_addr;
        server_addr.sin_family = AF_INET;
        server_addr.sin_addr.s_addr = INADDR_ANY;
        server_addr.sin_port = htons(port);
        
        if (bind(server_fd, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
            std::cerr << "❌ Failed to bind socket" << std::endl;
            close(server_fd);
            return 1;
        }
        
        // Listen for connections
        if (listen(server_fd, 5) < 0) {
            std::cerr << "❌ Failed to listen on socket" << std::endl;
            close(server_fd);
            return 1;
        }
        
        std::cout << "🚀 Noise-NK C++ server listening on localhost:" << port << std::endl;
        std::cout << "📡 Waiting for JavaScript clients..." << std::endl;
        
        while (server_running) {
            struct sockaddr_in client_addr;
            socklen_t client_len = sizeof(client_addr);
            
            int client_fd = accept(server_fd, (struct sockaddr*)&client_addr, &client_len);
            if (client_fd < 0) {
                if (server_running) {
                    std::cerr << "❌ Failed to accept connection" << std::endl;
                }
                continue;
            }
            
            // Get client address string
            char client_ip[INET_ADDRSTRLEN];
            inet_ntop(AF_INET, &client_addr.sin_addr, client_ip, INET_ADDRSTRLEN);
            std::string client_addr_str = std::string(client_ip) + ":" + std::to_string(ntohs(client_addr.sin_port));
            
            // Handle client in a separate thread
            std::thread client_thread(handle_client, client_fd, client_addr_str, server_private_key);
            client_thread.detach();
        }
        
        close(server_fd);
        std::cout << "🔌 Server stopped" << std::endl;
        
    } catch (const std::exception& e) {
        std::cerr << "❌ Server error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
} 